### ISSUE TITLE
External server causing strange HTTP_HOST behaviour in PHP

### ISSUE THREAD

**Issue by baldurpan** [bdkjones/CodeKit2#440]
Came across this weirdness yesterday at work and decided to try it at home, plus on a friends' server from as well, giving me the same confusing results.
Although, I'm not quite sure if this post really belongs here or if it's a PHP bug, but this behaviour doesn't seem to happen when testing with ssh port forwarding.

Anywho...

Basically, I the external server set to sprout.dev (a local apache+php setup) and my CodeKit server address on macbook-air.local:5757

Testing out this code in index.php:

``` php
<?php

var_dump($_SERVER['HTTP_HOST']);
var_dump(substr($_SERVER['HTTP_HOST'], -5, 5));

// Different version, Same results
var_dump($_SERVER['HTTP_HOST'], substr($_SERVER['HTTP_HOST'], -5, 5));

// Another version, set to a variable first. Same results
$host = $_SERVER['HTTP_HOST'];
var_dump($host, substr($host, -5, 5));
```

Gives me this result:

```
string(10) "macbook-air.local:5757"
string(5) "t.dev"
```

While the expected result would be:

```
string(22) "macbook-air.local:5757"
string(5) ":5757"
```

Note that the string(10) in the actual results seems to reflect the string length of the external server domain (sprout.dev) and not the actual string result (macbook-air.local:5757).

What do you think? Is CodeKit sending out some weird wibes to the web server or is this simply a strange bug in PHP?

**Comment by bdkjones** [comment]
You'd need to dump all the HTTP headers to be sure, but CodeKit sets the host header to your external server address before forwarding the request.

**Comment by baldurpan** [comment]
Debug output from getallheaders():

```
Host: [string]: macbook-air.local:5757
Accept: [string]: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Cache-Control: [string]: max-age=0, no-cache, no-store, must-revalidate
Via: [string]: macbook-air.local:5757
Connection: [string]: keep-alive
Cookie: [string]: sprout-dev=9giapql30nfg8esah3pikc2de2; ckrp_ypos=0_0
User-Agent: [string]: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2159.4 Safari/537.36
Accept-Language: [string]: en-US,en;q=0.8,is;q=0.6,lt;q=0.4
```

And from apache_response_headers():

```
X-Powered-By: [string]: PHP/5.5.17
Expires: [string]: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: [string]: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: [string]: no-cache
```

Also debug output from $_SERVER variable:

```
REDIRECT_STATUS: [string]: 200
HTTP_HOST: [string]: macbook-air.local:5757
HTTP_ACCEPT: [string]: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
HTTP_CACHE_CONTROL: [string]: max-age=0, no-cache, no-store, must-revalidate
HTTP_VIA: [string]: macbook-air.local:5757
HTTP_CONNECTION: [string]: keep-alive
HTTP_COOKIE: [string]: sprout-dev=9giapql30nfg8esah3pikc2de2; ckrp_ypos=0_0
HTTP_USER_AGENT: [string]: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2159.4 Safari/537.36
HTTP_ACCEPT_LANGUAGE: [string]: en-US,en;q=0.8,is;q=0.6,lt;q=0.4
HTTP_DNT: [string]: 1
HTTP_ACCEPT_ENCODING: [string]: 
PATH: [string]: /usr/bin:/bin:/usr/sbin:/sbin
SERVER_SIGNATURE: [string]: 
SERVER_SOFTWARE: [string]: Apache/2.2.26 (Unix) DAV/2 PHP/5.5.17 mod_ssl/2.2.26 OpenSSL/0.9.8za
SERVER_NAME: [string]: macbook-air.local:5757
SERVER_ADDR: [string]: 127.0.0.1
SERVER_PORT: [string]: 80
REMOTE_ADDR: [string]: 127.0.0.1
DOCUMENT_ROOT: [string]: /Users/baldur/Code/web/macbook-air.local:5757
SERVER_ADMIN: [string]: you@example.com
SCRIPT_FILENAME: [string]: /Users/baldur/Code/web/macbook-air.local:5757/index.php
REMOTE_PORT: [string]: 64705
REDIRECT_URL: [string]: /
GATEWAY_INTERFACE: [string]: CGI/1.1
SERVER_PROTOCOL: [string]: HTTP/1.1
REQUEST_METHOD: [string]: GET
QUERY_STRING: [string]: 
REQUEST_URI: [string]: /
SCRIPT_NAME: [string]: /index.php
PHP_SELF: [string]: /index.php
REQUEST_TIME_FLOAT: [double]: 1411291127.886
REQUEST_TIME: [integer]: 1411291127
```

Also strikes me as odd that the DOCUMENT_ROOT seems to be changing.

**Comment by bdkjones** [comment]
Interesting. 

All those headers reflect the changes that CodeKit's proxy server makes except for HOST, which should be the external server you set in the Project Settings within CodeKit.

At the time PHP runs, the HOST header should no longer be macbook-air.local:5757, which is the value of that header sent to CodeKit's proxy server from your browser during the initial HTTP request. CodeKit changes that header when it forwards the request to your external server. 

That change is reflected correctly when you tried logging the last five characters of the HOST header. I don't know why it's not reflected correctly in the dump above.

**Comment by baldurpan** [comment]
Well, this is a bit of a headache since I'm loading my system's config based on HTTP_HOST.
Is nobody else able to recreate this?

**Comment by baldurpan** [comment]
Not sure if it's relevant, but it might be worth noting that I'm using dnsmasq with this configuration

```
address=/.dev/127.0.0.1
```

**Comment by bdkjones** [comment]
Please set up a very small demo project with a single PHP file logging the headers, then send it to me so I can test over here. Thanks.

**Comment by bdkjones** [comment]
I have no idea why this behavior is happening, but here's hoping someone can figure it out --> http://stackoverflow.com/questions/25985204/php-http-host-value-incorrect-in-proxy-origin-server-setup

**Comment by bdkjones** [comment]
Got it. 

The Apache server IS working correctly and the Host header IS `localhost:8888` as it SHOULD be. 

However, when MAMP sends its response back to CodeKit, CodeKit replaces all instances of the string `localhost:8888` on the page with the string `vdk.local:5757` (or whatever your preview address is on your Mac). This is to keep links on the page referencing the proxy server rather than pointing directly at MAMP. 

That's why the browser output behaves the way it does --> When you log a substring of the HOST header, the string-replacement doesn't occur because CodeKit sees only the substring, not the full `localhost:8888`. 

In short, there's nothing wrong here --> MAMP has the correct Host header.

**Comment by baldurpan** [comment]
Classic trolling bug. :)
Thanks!
